### ISSUE TITLE
1-2% of vertices corrupted

### ISSUE THREAD

**Issue by bdeggleston** [thinkaurelius/titan#219]
Something is causing 'phantom' properties to be added to vertices.

1-2% of the vertices in our graph have a property on them which references a titan type that doesn't exist. This is causing all sorts of problems since it causes a null pointer exception to get thrown when getProperties is called on a vertex with this.

The null pointer exception is being thrown here:
https://github.com/thinkaurelius/titan/blob/titan02/titan-core/src/main/java/com/thinkaurelius/titan/graphdb/database/StandardTitanGraph.java#L360

I have a test that simulates the problem here:
https://github.com/StartTheShift/titan-tools/blob/master/src/test/java/com/shift/titantools/base/VertexRepairTest.java#L97

Some additional observations:

The contents of a few of the mystery Entry objects are below:
- col: [1,12,-110], val: [0,47,40,125,-113]
- col: [1,12,-110], val: [0,47,19,47,-11]
- col: [1,12,-110], val: [0,47,44,30,-117]
- col: [1,12,-110], val: [0,47,40,116,-103]

They appear to consistently be 3x5 bytes, none of the other properties on the vertex have that combo of col/val sizes. The only other property on the vertices that is consistently 5 bytes is #System#VertexState. I haven't seen any properties with 3x5 bytes on non corrupted vertices. Additionally, the corrupt vertices aren't missing any data, they just have an extra property on them.

The element id of the missing titan type is always 36028797018965522

We were thinking that this may be caused by the same type definitions being created by two Rexster servers simultaneously, with the losing machine caching a type that was never saved, but this problem survives restarting the Rexster server.

The vertices aren't being corrupted at an even rate. We don't have a ton of historical data, but it looks like vertices get corrupted for a few days, then there are 5 days of no corruption, then they start getting corrupted again. During the days that vertices are being corrupted, 0.5-2.5% of vertices updated that day get corrupted, but we have 2 days where the corruption rates were 18% and 11%

**Comment by bdeggleston** [comment]
so I have some updates on this, we ran cassandra's node repair tool, and things seem to have gotten much worse. I'm not sure if the two are related, but the timing would indicate it is. We now have vertex properties which point to 15 types that don't exist.

```
vertex check completed
  > 1549717 vertices examined
  > 29524 corrupt vertices detected
  > min timestamp: 1.362952850541876E9
  > max timestamp: 1.365786936876979E9
  > bad type ids:
    36028797018964542
    36028797018965042
    36028797018965010
    36028797018965522
    36028797018965118
    36028797018965018
    36028797018965530
    36028797018964946
    36028797018964706
    36028797018964674
    36028797018964558
    36028797018964998
    36028797018965098
    36028797018964550
    36028797018964990
```

Additionally, when running g.getProperties(), I'm getting a `The value is already used by another vertex and the key is unique` exception, thrown here: https://github.com/thinkaurelius/titan/blob/titan02/titan-core/src/main/java/com/thinkaurelius/titan/graphdb/transaction/AbstractTitanTx.java#L396

Any ideas?

**Comment by bdeggleston** [comment]
So this appears to be an edge case related to dropping keyspaces in cassandra, not a bug in Titan, though I haven't been able to duplicate

Before launching our application internally, we ran a script that would kill rexster, drop the titan keyspace via cqlsh, then start rexster. While this worked fine in our development environment, it seems like doing these operations in quick succession with a cassandra cluster resulted in data from the previous keyspace sticking around on some of the cassandra nodes.

Initially, this appeared to be a titan problem because existing vertices started accumulating the strange properties. Speculating that this might be the result of data inconsistencies between cassandra nodes, we ran node tool repair on the Cassandra cluster, hoping it would clean up these phantom properties. Not only did this not solve the problem, but it made it much worse, adding ~7000 "new" vertices that were missing properties that all of our vertices are supposed to have. However, we store a timestamp on each of our vertices, and this property was not missing for the most part. The timestamps for the new vertices were all from before the database wipe, going back about two weeks, which is what led us to believe that we were seeing old data that should have been removed. Since the application had not been launched, we took the opportunity to abandon the corrupted graph and point our application to a new rexster/titan cluster, and have not seen any problems since.

While trying to isolate and fix this problem, I wrote a utility that checks and cleans vertices and indices, which may be useful if you're having problems with corrupt vertices or indices. Link below:

https://github.com/StartTheShift/titan-tools
